### PR TITLE
Integrate the link picker

### LINKED ISSUE #2813: Link picker
Based on https://github.com/CollaboraOnline/online/pull/5852 we should:

- [x] Extend checkFileInfo to set the feature flag enabled
- [x] Listen for the post message to show the link picker dialog and emit the result to Collabora
- [x] Listen for the event to fetch the link preview metadata and emit the results to Collabora

### PR DESCRIPTION
**Pull request by julien-nc**
closes #2813

* [x] Extend checkFileInfo to set the feature flag enabled
* [x] Listen for the post message to show the link picker dialog and emit the result to Collabora
* [x] Listen for the event to fetch the link preview metadata and emit the results to Collabora

Tested without using a CODE server which supports the feature. Tested by reacting to `UI_Mention` messages. Worked fine :grin:.

Problems:
* [x] The link picker does not get the provider list even if `RenderReferenceEvent` is dispatched (I checked it really is). Could it be that a script is loaded too early?
* [x] I'm not so sure with the base64 encoding of the link thumbnail.
* [ ] Should we check the mimetype of the thumbnail request response? If so, we need to adjust `/core/references/preview/HASHED_REF` to provide a mimetype, for now, `application/octet-stream` is returned.

### CODE REVIEW

File: lib/Controller/WopiController.php [hunk 1]
```diff
@@ -223,6 +223,7 @@ public function checkFileInfo($fileId, $access_token) {
 			'DownloadAsPostMessage' => $wopi->getDirect(),
 			'SupportsLocks' => $this->lockManager->isLockProviderAvailable(),
 			'IsUserLocked' => $this->permissionManager->userIsFeatureLocked($wopi->getEditorUid()),
+			'EnableRemoteLinkPicker' => (bool)$wopi->getCanwrite(),
```
Comment (juliushaertl) [inline lib/Controller/WopiController.php:226]: I think the link picker should for now also be disabled for public links and direct editing (from the mobile clients)

File: src/document.js [hunk 2]
```diff
@@ -648,6 +656,43 @@ const documentsMain = {
...
+					if (imageResponse?.status === 200 && imageResponse?.data) {
+						b64Image = btoa(unescape(encodeURIComponent(imageResponse.data)))
+					}
+				} catch (e) {
+					console.error('Error loading the reference image', e)
+				}
+			}
+			PostMessages.sendWOPIPostMessage('loolframe', ' Action_GetLinkPreview_Resp', { url, title, image: b64Image })
+		} catch (e) {
+			console.error('Error resolving a reference', e)
```
Comment (juliushaertl) [inline src/document.js:692]: Can we show a toast also here so the user knows that something failed?

File: src/document.js [hunk 3]
```diff
@@ -648,6 +656,43 @@ const documentsMain = {
 		PostMessages.sendWOPIPostMessage('loolframe', 'Action_Mention', { list })
 	},
 
+	openLinkPicker() {
+		getLinkWithPicker(null, true)
```
Comment (juliushaertl) [inline src/document.js:660]: I'd also switch to async await here ;)

File: lib/Listener/LoadViewerListener.php [hunk 4]
```diff
@@ -56,6 +63,7 @@ public function handle(Event $event): void {
 		}
 		if ($this->permissionManager->isEnabledForUser() && $this->userId !== null) {
 			$this->initialStateService->provideCapabilities();
+			$this->eventDispatcher->dispatchTyped(new RenderReferenceEvent());
```
Comment (juliushaertl) [inline lib/Listener/LoadViewerListener.php:66]: document.js is loaded in an iframe already, so triggering the event rather ehe should work: https://github.com/nextcloud/richdocuments/blob/e3105a17e6b9f4baa4182f97d05f7b6f39e4d7a9/lib/Controller/DocumentTrait.php#L15

This is how the frames are currently structured, I've been thinking for a while if we shouldn't refactor that at some point but that may be larger work:

- Nextcloud loads viewer.js
  - The Office.vue component loads an iframe rendering a simple template through the DocumentController with document.js
    - document.js submits the wopi data to Collabora within another iframe

**Comment by juliushaertl** [comment]
> * Should we check the mimetype of the thumbnail request response? If so, we need to adjust `/core/references/preview/HASHED_REF` to provide a mimetype, for now, `application/octet-stream` is returned.

I'd say this is fine as we already validate before storing the image in https://github.com/nextcloud/server/blob/f0589523c80749b1bdfbbfea34b141cd562e216b/lib/private/Collaboration/Reference/LinkReferenceProvider.php#L163

File: lib/Controller/WopiController.php (hunk 1 above)
Comment (julien-nc) [reply lib/Controller/WopiController.php]: True, not sure about the new conditions.

File: src/document.js (hunk 2 above)
Comment (julien-nc) [reply src/document.js:694]: :+1:

File: src/document.js (hunk 3 above)
Comment (julien-nc) [reply src/document.js]: :+1:

File: lib/Listener/LoadViewerListener.php (hunk 4 above)
Comment (julien-nc) [reply lib/Listener/LoadViewerListener.php]: :heart:

**Comment by julien-nc** [comment]
@juliushaertl Do you mean detecting this without changing the link picker implementation? Just by checking if the picker result is a URL or not? Seems good to me.

**Comment by julien-nc** [comment]
@juliushaertl Done, it was fun to try it with the GPT-3 provider.

**Comment by juliushaertl** [comment]
Nice, will do a full test later with the collabora branch :)

File: src/document.js [hunk 5]
```diff
@@ -648,6 +657,53 @@ const documentsMain = {
...
+				try {
+					const imageResponse = await axios.get(thumbnailUrl)
+					if (imageResponse?.status === 200 && imageResponse?.data) {
+						b64Image = btoa(unescape(encodeURIComponent(imageResponse.data)))
+					}
+				} catch (e) {
+					console.error('Error loading the reference image', e)
+				}
+			}
+			PostMessages.sendWOPIPostMessage('loolframe', ' Action_GetLinkPreview_Resp', { url, title, image: b64Image })
```
Comment (eszkadev) [inline src/document.js:700]: There is a type here ' Action...' <- space as first character

**Comment by eszkadev** [comment]
what is expected thumbnail value format you send?
for me best would be something I can put in img.src without processing.
Like: `img.src = 'data:' + receivedImage`

**Comment by julien-nc** [comment]
@eszkadev Thanks for the review. There you go. It is now correctly requesting the image and getting a valid base64 string out of it.

**Review by juliushaertl — APPROVED** [review]
Tested and works. Thanks a lot @julien-nc

**Comment by eszkadev** [comment]
Thanks a lot! Tested and works with website URLs, but with picked link to the board from deck app I got `data:application/octet-stream;base64,PHN2ZyB4b...` and it cannot be displayed by the browser

**Comment by juliushaertl** [comment]
We'll have to address this separately in server

**Comment by julien-nc** [comment]
@juliushaertl Actually I found a fix for that. It's true our reference preview endpoint sometimes does not return the correct mimetype. But for images like described by @eszkadev, the response mimetype is actually correct. The issue is the way we build the Blob to produce the base64 string. I'll make another PR.

**Comment by pedropintosilva** [comment]
one more think, I noticed that typing a url with the protocol (http or https) doesn't work. Can we, when that is not entered by the user assume the protocol ?

**Pull request merged**